feat: Add streaming API to orchestration

tomfrenken opened the pull request:
## Context

SAP/ai-sdk-js-backlog#161.

This PR adds a streaming API to our orchestration client.

## Definition of Done

- [x] Code is tested (Unit, E2E)
- [x] Error handling created / updated & covered by the tests above
- [x] Documentation updated
- [x] (Optional) Aligned changes with the Java SDK
- [ ] (Optional) Release notes updated

ZhongpinWang commented on packages/foundation-models/src/azure-openai/azure-openai-embedding-client.ts:
```diff
@@ -9,7 +9,7 @@ import { apiVersion, type AzureOpenAiEmbeddingModel } from './model-types.js';
 
 /**
- * Azure OpenAI client for embeddings.
+ * Azure OpenAI client for embeddings./home/i519840/code/ai-sdk-js/packages/orchestration/src.
```
🚨🚨🚨

ZhongpinWang commented on packages/orchestration/src/orchestration-chat-completion-stream.ts:
```diff
@@ -0,0 +1,170 @@
+   * @internal
+   */
+  static async *_processFinishReason(
+    stream: OrchestrationChatCompletionStream<OrchestrationChatCompletionStreamChunkResponse>,
```
[pp] These long type names are eye catching 😆 But I guess it is what it is.

ZhongpinWang commented on packages/orchestration/src/orchestration-chat-completion-stream.ts:
```diff
@@ -0,0 +1,170 @@
+            if (response) {
+              response._getFinishReasons().set(choiceIndex, finishReason);
+            }
+            switch (finishReason) { // TODO: Cover all finish reasons
```
[q] Is there any other finish reason? Can you share this knowledge with me? Pretty curious.

tomfrenken replied:
I was under the assumption that other filters also write into the finish reason, but apparently they throw 400 instead.
I've removed the todo.

ZhongpinWang commented on packages/orchestration/src/orchestration-chat-completion-stream-chunk-response.ts:
```diff
@@ -0,0 +1,39 @@
+   * @param choiceIndex - The index of the choice to parse.
+   * @returns The finish reason.
+   */
+  getFinishReason(choiceIndex = 0): string | undefined {
```
[q] Do we know if orchestration plans to support multi-choice? I guess our solution is more generic, but just wanna be sure that this will not cause the chunk to look differently and we are still trying to access the chunks correctly.

ZhongpinWang commented on packages/orchestration/src/orchestration-client.ts:
```diff
@@ -50,17 +53,57 @@ export class OrchestrationClient {
+    prompt?: Prompt,
+    requestConfig?: CustomRequestConfig
+  ): Promise<OrchestrationChatCompletionStream<CompletionPostResponseStreaming>> {
+    const body = constructCompletionPostRequest(this.config, prompt, true);
+    const deploymentId = await resolveDeploymentId({
+      scenarioId: 'orchestration',
+      resourceGroup: this.deploymentConfig?.resourceGroup
+    });
+
+    const response = await executeRequest(
+      {
+        url: `/inference/deployments/${deploymentId}/completion`,
+        resourceGroup: this.deploymentConfig?.resourceGroup
+      },
+      body,
+      {
+        ...requestConfig,
+        signal: controller.signal
+      }
+    );
+    return OrchestrationChatCompletionStream._create(response, controller);
```
[pp] This part is a bit inconsistent to Azure OpenAI client. I think I tried to extract the common part out from both streaming and non-streaming variants to avoid code duplication like get deployment ID and construct `endpointOptions`.

tomfrenken replied:
Yeah that's a good idea, I'll create a common method.

ZhongpinWang commented on packages/orchestration/src/orchestration-client.ts:
```diff
@@ -50,17 +53,57 @@ export class OrchestrationClient {
 ): CompletionPostRequest {
   return {
     orchestration_config: {
+      stream,
```
[q] Do you know if orchestration also needs `stream_options` to include token usage in the last chunk?

```ts
stream_options: {
  include_usage: true
}
```

tomfrenken replied:
I didn't find anything in their docs.
They do have streaming options for specific modules, but didn't specify any usage related options.

Related docs: https://help.sap.com/docs/sap-ai-core/sap-ai-core-service-guide/streaming?locale=en-US

MatKuhr commented on sample-code/src/orchestration.ts:
```diff
@@ -39,6 +41,32 @@ export async function orchestrationChatCompletion(): Promise<OrchestrationRespon
 }
 
+/**
+ * Ask GPT 3.5 through the orchestration service about SAP Cloud SDK with streaming.
```
```suggestion
 * Ask GPT 4o through the orchestration service about SAP Cloud SDK with streaming.
```

tomfrenken replied:
I changed the endpoint to use 3.5 turbo instead (since that's cheaper)

MatKuhr commented on sample-code/src/orchestration.ts:
```diff
@@ -39,6 +41,32 @@ export async function orchestrationChatCompletion(): Promise<OrchestrationRespon
+    }
+  });
+
+  const response = await orchestrationClient.stream(undefined, controller);
```
currently, this API doesn't allow users to modify the stream options of orchestration. Eventually, this should be possible, so if not done with this BLI please create a follow-up so that we don't forget 😉

tomfrenken replied:
Also I checked with Dennis, if you provide streaming options, but call a non-streaming API, they are simply ignored.

MatKuhr replied:
The client expects `OrchestrationModuleConfig`, but it's not part of the module config, but on `orchestration_config` instead. So I wouldn't know how it can be customized currently..

> a non-streaming API
what is that?

tomfrenken replied:
> The client expects OrchestrationModuleConfig, but it's not part of the module config, but on orchestration_config instead. So I wouldn't know how it can be customized currently..

Yeah right, I just saw the options in the orchestration documentation, I would assume they end up in the spec at which point we will re-generate, then the config types will include the streaming options.

> what is that? Do you mean a model that doesn't support streaming?
No, rather if you send the config including streaming options but against a regular non-streaming chat-compltion API, e.g. you didn't set `streaming: true`

MatKuhr replied:
yeah, so with the current API we would either have to make it part of the prompt, or a new, separate parameter, or somehow else allow the user to pass this.

Okay, so you mean sending e.g.:
```json
{ "orchestration_config": 
  {
     "stream": false,
     "stream_options": {
         "chunk_size": 20
      }
  // ...
  }
}
```
?

tomfrenken replied:
Yes exactly.
You can still pass streaming options as part of the options object, but you would have to typecast it to "any", as an ugly workaround.

MatKuhr replied:
which options object? I don't see how currently the stream options can be passed at all..

tomfrenken replied:
In the initialization of the client.

MatKuhr replied:
I might be blind, but can you give an example code snippet? I don't see how this works..

tomfrenken replied:
```ts
const client = new OrchestrationClient(
  {
    llm: {
      model_name: 'gpt-35-turbo-16k',
      model_params: {}
    },
    templating: {
      template: [{ role: 'user', content: 'Create {number} paraphrases of {phrase}' }]
    },
    filtering: {
      input: {
        filters: [{ type: 'azure_content_safety', config: { Hate: 4, SelfHarm: 0 } }],
        streaming_options: { whatever: 'whatever' }
      } as InputFilteringConfig,
      streaming_options: {
          whatever: 'whatever'
      }
    } as FilteringModuleConfig
  }
);
```

MatKuhr commented on packages/orchestration/src/orchestration-client.ts:
```diff
@@ -85,3 +140,13 @@ export function constructCompletionPostRequest(
+  if (stream) {
+    llm.model_params = {
+      ...llm.model_params,
+      stream_options: { include_usage: true }
```
(question) is this API harmonized? Maybe test it with e.g. Gemini?

tomfrenken replied:
I was raising the same question, but at least also for Gemini it works as intended.

MatKuhr commented on packages/orchestration/src/orchestration-chat-completion-stream-chunk-response.ts:
```diff
@@ -0,0 +1,45 @@
+/**
+ * Orchestration chat completion stream chunk response.
+ */
+export class OrchestrationChatCompletionStreamChunkResponse {
```
Currently, this API is lacking the orchestration specific results, i.e. users can't access `module_results`

tomfrenken replied:
Yeah I had a convenience method to get the different module_results, but wasn't sure if anyone actually would use them.

You can still access them though, just not through a function, instead it's `chunk.data.module_results?.YOUR_MODULE`

tomfrenken replied:
What do you think, should there be an additional function for it?

Previously my method was `.getModuleResult(<select-module-with-auto-complete-here>)`

MatKuhr replied:
I would assume the module results would be useful on the overall response object? And there it's currently not available, right?

![image](https://github.com/user-attachments/assets/e4eae016-ba91-4d08-a1ac-0b200cd0e70a)

I guess it depends a bit on the specific results one is interested in. If we take a chat bot as example use case then I need the templating result + full LLM response at the end and store it in memory. Currently, I'd have to aggregate this manually across the chunks, right?

tomfrenken replied:
I see, so you would like to save all module_results in an array which is available on the response object?

MatKuhr replied:
yeah, but I would expect we aggregate over all chunks, so there is only one module_results outcome at the end that includes everything

tomfrenken replied:
I see, I suppose this needs a little bit of investigation to see which implementation (e.g. aggregation) would be the most suitable, I would suggest I create a follow-up ticket for this?

MatKuhr replied:
follow up ticket is also fine 👍🏻

tomfrenken replied:
https://github.com/SAP/ai-sdk-js-backlog/issues/200

MatKuhr commented on packages/orchestration/src/orchestration-chat-completion-stream.ts:
```diff
@@ -0,0 +1,179 @@
+import { createLogger } from '@sap-cloud-sdk/util';
+import { SseStream } from '@sap-ai-sdk/core';
```
this file is mostly code duplication, and since this is not trivial code it would be nice to reduce it. I know the generics are annoying, but maybe you can find a way to represent them without nesting them deeper?

E.g. something alone those lines:
```
static async *_processChunk<T extends SseStream<T>>(
    stream: T
  ): AsyncGenerator<T> {
```

tomfrenken replied:
I suppose this could be done for most process functions, but not for 
`_processFinishReason()`, as this method accesses different properties based on the chunk type

tomfrenken replied:
https://github.com/SAP/ai-sdk-js-backlog/issues/199

ZhongpinWang commented on sample-code/src/orchestration.ts:
```diff
@@ -39,6 +41,37 @@ export async function orchestrationChatCompletion(): Promise<OrchestrationRespon
+      template: [
+        {
+          role: 'user',
+          content: 'Give me a short introduction of SAP Cloud SDK.'
```
Yes, but we use it in our e2e. And asking for a long description was the issue last time consuming too many tokens.

tomfrenken replied:
While true, I would like this to be used in our E2E tests in the future, do you think making the response longer is worth it still?

tomfrenken commented:
The biggest change of the additional refactor was, that I had to make most underlying functions static.

I've also added a few interfaces which simplify the overall API, so that you don't have to call the functions with a myriad of `undefined` before actually passing the parameters.

deekshas8 commented on packages/orchestration/src/orchestration-client.ts:
```diff
@@ -1,63 +1,177 @@
+    destination?: HttpDestinationOrFetchOptions,
+    requestConfig?: CustomRequestConfig
+  ): Promise<
+    OrchestrationChatCompletionStreamResponse<OrchestrationChatCompletionStreamChunkResponse>
```
`OrchestrationChatCompletionStreamResponse<OrchestrationChatCompletionStreamChunkResponse>`

Is way too long

deekshas8 commented on packages/orchestration/src/orchestration-client.ts:
```diff
@@ -1,63 +1,177 @@
+    return new OrchestrationResponse(response);
+  }
+
+  static async streamWithJson(
```
[req] Both static functions have too many parameters, maybe consider using object parameters instead.

ZhongpinWang replied:
I also feel it requires too many params. I am wondering, if it is really a bad idea to accept json in the constructor.

Consider the following cases (please verify if my understanding is correct):

| `stream` property in JSON | Which method to call | Consequence |
|---|---|---|
| `true` | `chatCompletion()` | All good, non-streaming with `stream` overwritten to `false` |
| `true` | `stream()` | All good, streaming |
| `false` | `chatCompletion()` | All good, non-streaming |
| `false` | `stream()` | All good (?), streaming with `stream` overwritten to `true` |

KavithaSiva commented on packages/orchestration/src/orchestration-chat-completion-stream-response.ts:
```diff
@@ -0,0 +1,57 @@
+    this._usage = usage;
+  }
+
+  public getFinishReason(choiceIndex = 0): string | undefined | null {
```
[req] I didn't see an example where `finish_reason` is `null`. I think if it is present it always has atleast an empty string.

```suggestion
  public getFinishReason(choiceIndex = 0): string | undefined {
```

KavithaSiva commented on packages/orchestration/src/orchestration-client.ts:
```diff
@@ -78,30 +192,47 @@ export function constructCompletionPostRequestFromJson(
-          grounding_module_config: config.grounding
-        })
-      }
+      stream,
```
[pp] 
```suggestion
      ...(stream && { stream: true })
```

tomfrenken replied:
While this somewhat aligns with what we did with the other properties, I disagree here.
This convolutes the logic of our code, even though it is not a requirement of the target API.

If anything, I think orchestration should be able to handle empty configs for the different modules, so that we can simply forward it without doing the whole spread logic.

KavithaSiva replied:
I felt spread logic is helpful, because then you wouldn't have to modify the existing tests to add `stream: false` for a non-streaming variant request.

And moreover, since `stream_options` can also be added here, I think spreading is still better. WDY think?

tomfrenken replied:
I've created a separate utility function to add / remove streaming options, it will still set streaming to false, but this is by design, as people could instantiate a client with a streaming module config, but call chat completion, in which case we need to overwrite the value to false, or delete it entirely. Since deleting the field seems to be too much additional logic for me, I set it to false for now.

KavithaSiva commented on packages/orchestration/src/orchestration-chat-completion-stream.ts:
```diff
@@ -0,0 +1,179 @@
+                response._getFinishReasons().set(choiceIndex, finishReason);
+              }
+              switch (finishReason) {
+                case 'content_filter':
```
[q] Any idea, if we can have more finish reasons here other than the ones similar to Open AI?

tomfrenken replied:
AFAIK no.

KavithaSiva reviewed (changes requested):
The stream functionality itself looks mostly good, but, I would also consider being able to provide streaming options like `chunk_size` before merging though.

tomfrenken commented on packages/orchestration/src/orchestration-chat-completion-stream-response.ts:
```diff
@@ -0,0 +1,57 @@
+/**
+ * Orchestration chat completion stream response.
+ */
+export class OrchestrationChatCompletionStreamResponse<T> {
```
Yeah agreed, I aligned with the existing streaming implementation here, but I suppose this would be better naming wise.

tomfrenken commented on packages/orchestration/src/orchestration-client.ts:
```diff
@@ -1,63 +1,177 @@
+    return new OrchestrationResponse(response);
+  }
+
+  static async streamWithJson(
```
I disagree, the call we make is exactly the same as we do with our existing streaming implementation, and as the config is in the user's hands, they are responsible for providing a valid config.

tomfrenken replied:
All the additional test really would do is validating that JSON.parse() works.

tomfrenken commented on packages/orchestration/src/orchestration-chat-completion-stream-chunk-response.test.ts:
```diff
@@ -0,0 +1,79 @@
+        'orchestration',
+        'orchestration-chat-completion-stream-chunk-response-token-usage.json'
+      ),
+      finishReasonResponse: await parseMockResponse<any>(
+        'orchestration',
+        'orchestration-chat-completion-stream-chunk-response-finish-reason.json'
+      ),
+      deltaContentResponse: await parseMockResponse<any>(
+        'orchestration',
+        'orchestration-chat-completion-stream-chunk-response-delta-content.json'
```
While true, I wasn't sure if merging them is the best approach, as they are potentially separate.
E.g., in this case, there was no real finish reason, since it was a successful stream, however, if the stream stops due to a filter hit, the token usage wouldn't be part of the same chunk, and even be omitted AFAIK.

I'm fine for merging these for the test, but what do you think?

KavithaSiva replied:
Apologies, I meant `orchestration-chat-completion-stream-chunk-response-token-usage.json` and `test-util/data/orchestration/orchestration-chat-completion-stream-chunk-response-finish-reason.json` as they both are essentially same.

So, you can just use one file for both.
`orchestration-chat-completion-stream-chunk-response-delta-content.json` is a different output, so you can still leave that one in.

ZhongpinWang commented on packages/orchestration/src/orchestration-client.ts:
```diff
@@ -1,63 +1,177 @@
     return new OrchestrationResponse(response);
   }
+
+  async stream(
```
I feel currently `stream_options` is not supported at all?

KavithaSiva replied:
To add it to filter module, we can still use for e.g :
```TS
const filtering: FilteringModuleConfig = {
  input: buildAzureContentFilter(),
  output: {
    ...buildAzureContentFilter(),
    ...{
      stream_options: {
        overlap: 10
      }
    }
  }
};

```
as the `OutputFilteringConfig` has the streaming options already generated in.
but I don't think there is an easy way yet to add `stream_options` with `chunk_size`

tomfrenken commented:
Alright, I've made the suggested changes to all of the logic, I still need to add some additional tests though, in the meantime, you can check if my changes align with your suggestions.

Changes:
- API has been shortened, as multiple parameters where put into a single options object
- You can pass streaming options now
- Renamed all functions to be shorter
- Renamed JSON functions for clarity

ZhongpinWang commented on packages/orchestration/src/orchestration-client.ts:
```diff
@@ -78,31 +202,73 @@ export function constructCompletionPostRequestFromJson(
+        }
+      })
+    }
+  } as OrchestrationConfig; // TODO: Remove typecast when types are re-generated;
```
[q] Can we fix this within this PR? I see some type issues if not casting it.

deekshas8 replied:
[req] I think you probably need to add a type check for output as well so that filters does not evaluate to undefined. 

```
...(moduleConfigs.filtering_module_config?.output && {
            output: {
            filters: moduleConfigs.filtering_module_config.output.filters,
            stream_options: outputFiltering
          }})
```

I am guessing there is a better way to write this.

ZhongpinWang commented on packages/orchestration/src/orchestration-client.ts:
```diff
@@ -78,31 +202,73 @@ export function constructCompletionPostRequestFromJson(
+  };
+
+  return {
+    orchestration_config: stream
+      ? addStreamOptions(moduleConfigurations, streamOptions)
```
[pp] As previously mentioned, why don't we allow streamOptions to be optional and get rid of the stream flag.

ZhongpinWang reviewed (changes requested):
Review on going...

tomfrenken commented on packages/orchestration/src/orchestration-client.ts:
```diff
@@ -1,69 +1,193 @@
+}
+
+interface StreamOptions {
+  chunkSize?: number;
+  llm?: { includeUsage?: boolean; [key: string]: any };
+  // Add more options as they are implemented
+  // masking?: { };
+  // grounding?: { };
+  // inputFiltering?: { };
+  outputFiltering?: { overlap?: number };
```
Yeah makes sense to me, however, until the spec is updated (probably in 1.6?) these types don't exist yet.

ZhongpinWang replied:
At least `GlobalStreamOptions` and `FilteringStreamOptions` already exist. Should we just use those generated types and replace llm later once we have a newer spec?

tomfrenken replied:
Nvm, I mixed this up with the Azure types.

tomfrenken commented on packages/orchestration/src/orchestration-client.ts:
```diff
@@ -78,31 +202,73 @@ export function constructCompletionPostRequestFromJson(
+  moduleConfigs: ModuleConfigs,
+  streamOptions?: StreamOptions
+): OrchestrationConfig {
+  const llm = streamOptions?.llm ?? {};
```
Yeah good point, I'll rename it.
I'd rather not inline it, to keep the code a little more readable.

MatKuhr reviewed (commented):
in progress ...

deekshas8 commented on packages/orchestration/src/orchestration-client.ts:
```diff
@@ -1,69 +1,193 @@
+      config
+    } = options;
+
+    const body = jsonModuleConfig
```
[req] I had a similar preference when looking at this. I wouldn't create a completely separate `RequestOptions` type, but atleast have `config: string | OrchestrationModuleConfig` type and then add a type guard here so we avoid the `!` in `constructCompletionPostRequest(config!, prompt, stream, streamOptions);`

tomfrenken replied:
I narrowed the type down, do you still prefer a different approach?
A benefit of it is that I can use the spread syntax later on

deekshas8 commented on packages/orchestration/src/orchestration-client.ts:
```diff
@@ -1,69 +1,193 @@
+          jsonModuleConfig,
+          prompt
+        )
+      : constructCompletionPostRequest(config!, prompt, stream, streamOptions);
```
[q] Do we even need a `stream` bool property here? Isn't it enough to check if streamOptions was passed and has properties?

tomfrenken commented on packages/core/src/stream/line-decoder.ts:
```diff
@@ -5,7 +5,6 @@ type Bytes = string | ArrayBuffer | Uint8Array | Buffer | null | undefined;
  * reading lines from text.
  *
  * Https://github.com/encode/httpx/blob/920333ea98118e9cf617f246905d7b202510941c/httpx/_decoders.py#L258.
- * @internal
```
Cause you can't export internal functions, and everything in core is by definition "internal" anyway.

tomfrenken replied:
Well, that "public api" is literaly every exported function of the package.

tomfrenken replied:
We can ofc export it and reach it from the internal.ts path, but conceptually the entire core package is internal.

deekshas8 commented on packages/orchestration/src/orchestration-client.ts:
```diff
@@ -1,69 +1,193 @@
 
+interface RequestOptions {
+  prompt?: Prompt;
+  requestConfig?: CustomRequestConfig;
```
[pp] Please reuse type `ConfigurationOptions` here so we avoid duplication.

ZhongpinWang reviewed (changes requested):
I have gone through all the files once. Maybe there will be a second round review. 🔁

ZhongpinWang commented on packages/orchestration/src/index.ts:
The following missing exports are related to this PR:

- `ChatDelta`
- `CompletionPostResponseStreaming`
- `FilteringStreamOptions`
- `GlobalStreamOptions`
- `LLMChoiceStreaming`
- `LLMModuleResultStreaming`
- `LLMModuleResultSynchronous`

The following missing exports are template related. 
@MatKuhr I see that we export `TemplatingModuleConfig` ( `= Template | TemplateRef`) without exporting the following types:

- `TemplateRefByID`
- `TemplateRefByScenarioNameVersion`
- `TemplateRef`
- `Template`
- `TemplatingChatMessage`

ZhongpinWang commented on packages/orchestration/src/orchestration-client.ts:
```diff
@@ -1,113 +1,136 @@
+  ): Promise<OrchestrationStreamResponse<OrchestrationStreamChunkResponse>> {
+    if (typeof this.config === 'string' && options) {
+      logger.warn(
+        'Stream options are not supported when using a JSON module config.'
```
😄 Please if anyone has better idea than this, put a comment.

MatKuhr replied:
Some ideas: #427 427

1. Technically, we could narrow this condition to `options?.llm`.
2. Remove the separate handling for `string` and operate only on the object. 
  * Shorter code, but requires either casting to any or other shenanigans.
  * Mainly depends if we want to guarantee arbitrary payload works (assuming the server accepts it of course). 
  * Alternatively, we can be slightly more strict and assume some fields (essentially only `llm_module_config` currently) are always present.

tomfrenken replied:
I thought about a similar approach, but casting usually ends up masking changes to the underlying types throughout the package's lifecycle, especially when re-generating the types, which could end up causing issues down the line, therefore I would always try to avoid it.

Just like in your example, the only way to handle it like an arbitrary object is to make the type a resursivePartial, which would be an object that doesn't satisfy the current API (e.g., if you set any filter config it has to have N properties).

We could consider refactoring this in a follow-up PR, but honestly I would argue that if you are using a JSON config client, all the options that you want to have enabled should already be set at initialization of the client.

ZhongpinWang commented on packages/orchestration/src/orchestration-client.ts:
```diff
@@ -1,113 +1,136 @@
     const deploymentId = await resolveDeploymentId({
       scenarioId: 'orchestration',
-      resourceGroup: this.deploymentConfig?.resourceGroup
+      ...(this.deploymentConfig || {})
```
[pp] I would prefer using `??`, which sounds more correct to me, checking undefined or null instead of falsy. Same to other occurrences.

ZhongpinWang commented on packages/orchestration/src/orchestration-utils.ts:
Love this.

ZhongpinWang commented on packages/orchestration/src/orchestration-types.ts:
```diff
@@ -56,3 +59,39 @@ export interface OrchestrationModuleConfig {
+/**
+ * Options for the stream.
+ */
+export interface StreamOptions extends GlobalStreamOptions {
```
[pp] Or what about having an additional attribute called `global` to store chunk_size etc. Just in case in the future we have more and more, and we need to manually extract the global options one by one.

tomfrenken replied:
Yes!!

ZhongpinWang commented on packages/orchestration/src/orchestration-utils.ts:
```diff
@@ -0,0 +1,155 @@
+  return {
+    stream: true,
+    stream_options: {
+      chunk_size: chunkSize
```
See my previous comment, would be nice if we can do `stream_options: ...streamOptions?.global`

ZhongpinWang reviewed (commented):
One more round

ZhongpinWang commented on test-util/data/orchestration/orchestration-chat-completion-stream-chunk-response-delta-content.json:
```diff
@@ -0,0 +1,39 @@
+                    "role": "assistant",
+                    "content": "rimarily focusing on Java and JavaScript/Node.js environments, allowing developers to work in their "
+                },
+                "finish_reason": ""
```
Nvm ignore this, we already implemented a falsy check so "" is ignored. I was a bit confused.

MatKuhr commented on packages/orchestration/src/orchestration-stream-chunk-response.ts:
```diff
@@ -0,0 +1,48 @@
+   * @param choiceIndex - The index of the choice to parse.
+   * @returns The message delta content.
+   */
+  getDeltaContent(choiceIndex = 0): string | undefined {
```
Do we have tests for n > 1? If not, I'm not sure this really works as intended, if it depends on LLMs sending only one choice per chunk (which isn't guaranteed), and how users could meaningfully use this.

If we are unsure, we could also disable / remove it for now

tomfrenken replied:
Not sure if I understand the issue, can you elaborate?

The given function could handle a "out of bounds" choiceIndex, if a given LLM has more than one choice, it will either be found and returned, or undefined is returned, either would be fine AFAIK?

MatKuhr reviewed (commented):
Didn't get through everything, but some ideas. Also, let's remember to add E2E tests

ZhongpinWang commented on packages/orchestration/src/orchestration-client.ts:
```diff
@@ -1,69 +1,193 @@
+  async stream(
+    prompt?: Prompt,
+    controller = new AbortController(),
+    options?: StreamOptions,
```
We discussed this and feel it makes sense for user to use controller if they intend to use streaming in a correct way. Because without a controller, an ongoing stream can never be stopped in case of any error, or if user leave/refresh the page, connection closed, etc. So kind like a must have in the end. Worst case scenario, a new AbortController() is also not super complicated.

ZhongpinWang commented on packages/orchestration/README.md:
```diff
@@ -85,6 +85,124 @@ The client allows you to combine various modules, such as templating and content
+
+#### Stream Options
+
+The orchestration service offers multiple streaming options, which you can configure in addition to the LLM's streaming options.
```
[pp] multiple streaming options such as?

ZhongpinWang commented on packages/orchestration/README.md:
```diff
@@ -85,6 +85,124 @@ The client allows you to combine various modules, such as templating and content
+#### Stream Options
+
+The orchestration service offers multiple streaming options, which you can configure in addition to the LLM's streaming options.
+There are two ways to add specific streaming options to your client, either at initalization, or dynamically when calling the stream API.
```
```suggestion
There are two ways to add specific streaming options to your client, either at initialization of orchestration client, or when calling the stream API.
```

ZhongpinWang commented on packages/orchestration/README.md:
```diff
@@ -85,6 +85,124 @@ The client allows you to combine various modules, such as templating and content
+The orchestration service offers multiple streaming options, which you can configure in addition to the LLM's streaming options.
+There are two ways to add specific streaming options to your client, either at initalization, or dynamically when calling the stream API.
+
+Dynamically setting these options after client initialization is particularly helpful when you've initialized a client with a config meant for regular chat completion and now want to switch to using streaming.
```
```suggestion
Setting streaming options dynamically could be useful if an initialized orchestration client will also be used for streaming.
```

But I mean even if a user put stream options when initializing the client, they can still use the client for non-streaming chat completion right? I can only imagine setting stream options dynamically if user write some kind of playground or reuse the same orchestration client for different use cases (which requires e.g. different filter settings?)

ZhongpinWang commented on packages/orchestration/README.md:
````diff
@@ -85,6 +85,124 @@ The client allows you to combine various modules, such as templating and content
+);
+```
+
+Usage metrics are collected by default, if you do not want to receive them, set include_usage to false.
````
```suggestion
Usage metrics are returned by default, if you do not want to receive them, set `include_usage` to `false`.
```

ZhongpinWang commented on packages/orchestration/README.md:
````diff
@@ -85,6 +85,124 @@ The client allows you to combine various modules, such as templating and content
+```
+
+Usage metrics are collected by default, if you do not want to receive them, set include_usage to false.
+If you don't want any streaming options as part of your call to the LLM, set options.llm = null.
````
```suggestion
If you don't want any streaming options as part of your call to the LLM, set `streamOptions.llm` to `null`.
```

tomfrenken commented on packages/orchestration/src/orchestration-chat-completion-stream-chunk-response.test.ts:
```diff
@@ -0,0 +1,79 @@
+    };
+  });
+
+  it('should return the chat completion stream chunk response', () => {
```
We need 3 responses to replicate the "real" behavior of the orchestration service.
The reason we test on the chunk level instead of the stream.respones level is because I tried to test the lowest component / shift left approach.
We also do the same in the existing streaming tests.

tomfrenken commented on packages/orchestration/src/orchestration-client.test.ts:
```diff
@@ -378,4 +380,50 @@ describe('orchestration service client', () => {
+    );
+
+    for await (const chunk of response.stream) {
+      expect(chunk.data).toEqual(JSON.parse(initialResponse));
```
There are multiple chunks, however, only the first chunk is tested, which should match the expected initial response.
We could extend this to the entire stream, do you think that would be necessary?

ZhongpinWang commented on packages/orchestration/README.md:
```diff
@@ -86,6 +86,125 @@ The client allows you to combine various modules, such as templating and content
+  {
+    llm: { include_usage: false },
+    global: { chunk_size: 10 },
+    outputFiltering: { overlap: 200 }
```
[q] Out of curiosity, have you tried setting streaming options for filtering without providing a filter?

tomfrenken replied:
I did 👍

tomfrenken commented:
Side node while I add streaming to our sample code and e2e tests, we should really refactor our server.ts file, lots of duplication and it should be more modularized ...

ZhongpinWang commented on tests/e2e-tests/src/orchestration.test.ts:
```diff
@@ -62,4 +62,7 @@ describe('orchestration', () => {
     expect(response.getContent()?.includes('logo')).toBe(true);
   });
+
+  // add e2e test
+  // java tests only that there is any response
```
[q] TODO?

ZhongpinWang reviewed (changes requested):
After e2e added

ZhongpinWang commented on sample-code/README.md:
```diff
@@ -162,6 +162,20 @@ Use `buildAzureContentFilter()` to build the content filter.
 
+#### Chat Completion Streaming
+
+`GET /orchestration-stream/chat-completion-stream`
```
```suggestion
`POST /orchestration-stream/chat-completion-stream`
```
[req] Also add a payload example here. Would be nice to have to have a curl cmd here.

ZhongpinWang replied:
[req] `/orchestration-stream/chat-completion-stream-json` missing.

ZhongpinWang commented on .changeset/hip-melons-destroy.md:
```diff
@@ -0,0 +1,5 @@
+'@sap-ai-sdk/orchestration': minor
+---
+
+[New Funcionality] Add support for streaming in the orchestration client.
```
```suggestion
[New Functionality] Add support for streaming in the orchestration client.
```

ZhongpinWang reviewed (approved):
LGTM 🎉
